pkg/go/graph/weighted_graph.go

```diff
@@ -57,6 +57,19 @@ func (wg *WeightedAuthorizationModelGraph) AddNode(uniqueLabel, label string, no
 	wg.nodes[uniqueLabel] = &WeightedAuthorizationModelNode{uniqueLabel: uniqueLabel, label: label, nodeType: nodeType, wildcards: wildcards}
 }
 
+// AddNode adds a node to the graph with optional nodeType and weight.
+func (wg *WeightedAuthorizationModelGraph) GetOrAddNode(uniqueLabel, label string, nodeType NodeType) *WeightedAuthorizationModelNode {
+	if existingNode := wg.nodes[uniqueLabel]; existingNode != nil {
+		return existingNode
+	}
+	wildcards := make([]string, 0)
+	if nodeType == SpecificTypeWildcard {
+		wildcards = append(wildcards, uniqueLabel[:len(uniqueLabel)-2])
+	}
+	wg.nodes[uniqueLabel] = &WeightedAuthorizationModelNode{uniqueLabel: uniqueLabel, label: label, nodeType: nodeType, wildcards: wildcards}
```

Comment:
Looking at how the `wildcards` property is used throughout the `WeightedAuthorizationModelGraph`, you can avoid many unnecessary allocations by leaving `wildcards` equal to `nil` when there are no wildcard values present.
```suggestion
	var wildcards []string
	if nodeType == SpecificTypeWildcard {
		wildcards = append(wildcards, uniqueLabel[:len(uniqueLabel)-2])
	}
	wg.nodes[uniqueLabel] = &WeightedAuthorizationModelNode{uniqueLabel: uniqueLabel, label: label, nodeType: nodeType, wildcards: wildcards}
```

Reply:
Actually, thinking it a little bit more, we don't need the append at all as we already know that wildcards is an empty array
```
		var wildcards []string
		if nodeType == SpecificTypeWildcard {
			wildcards = []string{ uniqueLabel[:len(uniqueLabel)-2]}
		}
		wg.nodes[uniqueLabel] = &WeightedAuthorizationModelNode{uniqueLabel: uniqueLabel, label: label, nodeType: nodeType, wildcards: wildcards}
```

Reply:
Using append, or creating the array outright are both acceptable here; I'm not certain either would yield a significant performance benefit in this particular case.